kompile: add check for token sort productions

SchmErik opened the pull request:
Token sorts should only contain productions that have `token` or
`function` labels. Any other label (including subsorts) are illegal. Add
a check to ensure that token definitions are correct.

Fixes: #2119

daejunpark commented:
This PR will be subsumed by #2217.

@theo25 could you please check if #2217 (or keq branch 6e40fb5f0ec35925aa1ef387636690e7cedcad9e)  fixed your problem?

theo25 commented:
@daejunpark No, the problem is still there when trying with this branch/commit.

daejunpark commented:
@theo25 oh, sorry I forgot to say that you should use another command. `Kapi.main` is deprecated and will be deleted. Please use this command:
```
$ keq <prelude.smt2> <def0.k> <mod0> <def1.k> <mod1> <def2.k> <mod2> <spec.k>
```
For example,
```
$ cd k-distribution/tests/equiv/test2
$ keq ../../../../k-distribution/include/z3/basic.smt2 imp.k IMP imp1.k IMP imp2.k IMP if_spec.k
```
You can find more examples in https://github.com/kframework/k/blob/keq/k-distribution/tests/equiv/test2/run.sh

theo25 commented:
@daejunpark I am actually using the org.kframework.main.Main class from the shell module with -keq as first argument and the rest of the arguments as you suggested. Is this also deprecated?

daejunpark commented:
@theo25 you're right. Now I see what's the problem. I'll incorporate your fix into the PR.

daejunpark commented:
@theo25 could you please check if #2225 fixed the original problem?

daejunpark commented:
closed via #2225

Commit:
kompile: add check for token sort productions

Token sorts should only contain productions that have `token` or
`function` labels. Any other label (including subsorts) are illegal. Add
a check to ensure that token definitions are correct.

radumereuta commented on kore/src/main/scala/org/kframework/definition/outer-ext.scala:
```diff
@@ -64,6 +64,7 @@ object FlatModule {
         val newImports = m.imports.map(f)
         val newM = new Module(m.name, newImports, m.localSentences, m.att)
         newM.checkSorts()
+        newM.checkTokenSorts()
```
There are two locations where we call `checkSorts()`. Why are we calling `checkTokenSorts()` only once?

SchmErik replied:
Oh right, thanks for the reminder. I'll extract these two lines in a new function and place that in both places that call checkSorts()

radumereuta commented on kore/src/main/scala/org/kframework/definition/outer.scala:
```diff
@@ -363,6 +363,21 @@ case class Module(val name: String, val imports: Set[Import], localSentences: Se
+    val localTokens = localProductions
+      .collect({ case p if p.att.contains("token") => p })
+      .groupBy(_.sort)
+      .map { case (s, ps) => (s, ps) }
```
This can be compacted further. Haven't tested it but this might be equivalent:
```
localProductions.filter(p => p.att.contains("token")).groupBy(_.sort)
```

radumereuta commented on kore/src/main/scala/org/kframework/definition/outer.scala:
```diff
@@ -363,6 +363,21 @@ case class Module(val name: String, val imports: Set[Import], localSentences: Se
+      case p: Production => throw KEMException.compilerError(
+        "Sort "+ p.sort.name + " was declared as a token. Productions of this sort can only contain [function] or [token] labels.", p)
+      case _ =>
+    }
```
This can also be compacted further.
```
productions
.filter(c1 & c2)
.map(p => throw)
```

radumereuta reviewed (changes requested):
This could use a test.

SchmErik commented:
What's the expected behavior in this case?
```
  syntax X ::= "a"
  syntax X ::= "b" [token]
```
This current PR, this results in an error on these two lines

Commit:
check local productions against all tokens and exceptions sorts to avoid analyzing some predefined variables

SchmErik commented:
I made several large modifications to my solution so I'm going to close this and open a different pull request.
